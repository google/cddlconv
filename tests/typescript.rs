use std::io::BufWriter;

use cddl::visitor::Visitor;
use cddlconv;

macro_rules! test {
    ($name:ident, $input:expr) => {
        #[test]
        fn $name() {
            let input = std::fs::read_to_string($input).unwrap();
            let cddl = cddl::parser::cddl_from_str(&input, true).unwrap();
            let stdout = BufWriter::new(Vec::new());
            let stderr = BufWriter::new(Vec::new());
            let mut engine = cddlconv::engines::typescript::Engine::with_writers(stdout, stderr);
            engine.visit_cddl(&cddl).unwrap();
            engine.print_postamble();

            let (stdout, stderr) = engine.into_writers();
            insta::assert_snapshot!(String::from_utf8(stderr.into_inner().unwrap()).unwrap());
            insta::assert_snapshot!(String::from_utf8(stdout.into_inner().unwrap()).unwrap());
        }
    };
}

test!(it_works, "examples/webdriver-bidi/webdriver-bidi.cddl");
test!(it_works_with_arrays, "examples/rfc-examples/arrays.cddl");
test!(it_works_with_maps, "examples/rfc-examples/maps.cddl");
test!(
    it_works_with_amendments,
    "examples/rfc-examples/colors.cddl"
);
test!(it_works_with_optional_groups, "examples/optional_groups.cddl");
test!(it_works_with_simple_optional_groups, "examples/simple_optional_groups.cddl");
test!(it_works_with_array_occurences, "examples/array_occurences.cddl");
