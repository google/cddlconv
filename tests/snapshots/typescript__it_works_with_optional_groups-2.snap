---
source: tests/typescript.rs
assertion_line: 32
expression: "String::from_utf8(stdout.into_inner().unwrap()).unwrap()"
---
export type A = (({
"foo":(string)}&({} |({
"bar":(string)}))));
export type B = (({
"foo":(string)}&({} |Bar)));
export type Bar = ({
"bar":(string)});

